satpy/readers/clavrx.py

```diff
@@ -123,28 +149,31 @@ def _get_data(data, dataset_id: dict) -> xr.DataArray:
 
         attrs = data.attrs.copy()
 
-        fill = attrs.get("_FillValue")
+        # don't need these attributes after applied.
         factor = attrs.pop("scale_factor", (np.ones(1, dtype=data.dtype))[0])
         offset = attrs.pop("add_offset", (np.zeros(1, dtype=data.dtype))[0])
+        flag_values = data.attrs.get("flag_values", [None])
         valid_range = attrs.get("valid_range", [None])
-        if isinstance(valid_range, np.ndarray):
-            attrs["valid_range"] = valid_range.tolist()
 
-        flags = not data.attrs.get("SCALED", 1) and any(data.attrs.get("flag_values", [None]))
-        if not flags:
+        if isinstance(valid_range, np.ndarray):
+            valid_range = valid_range.tolist()
+            attrs["valid_range"] = valid_range
+
+        flags = not data.attrs.get("SCALED", 1) and any(flag_values)
+        if flags:
+            fill = attrs.get("_FillValue", None)
+            if isinstance(flag_values, np.ndarray) or isinstance(flag_values, list):
+                data = data.where((data >= flag_values[0]) & (data <= flag_values[-1]), fill)
+        else:
+            fill = attrs.pop("_FillValue", None)
             data = data.where(data != fill)
-            data = _CLAVRxHelper._scale_data(data, factor, offset)
-            # don't need _FillValue if it has been applied.
-            attrs.pop("_FillValue", None)
-
-        if all(valid_range):
-            valid_min = _CLAVRxHelper._scale_data(valid_range[0], factor, offset)
-            valid_max = _CLAVRxHelper._scale_data(valid_range[1], factor, offset)
-            if flags:
-                data = data.where((data >= valid_min) & (data <= valid_max), fill)
-            else:
+            data = _scale_data(data, factor, offset)
+
+            if valid_range[0] is not None:
+                valid_min = _scale_data(valid_range[0], factor, offset)
+                valid_max = _scale_data(valid_range[1], factor, offset)
                 data = data.where((data >= valid_min) & (data <= valid_max))
-            attrs["valid_range"] = [valid_min, valid_max]
+                attrs["valid_range"] = [valid_min, valid_max]
```

Comment:
I don't need this to be cleaned up before merging, but this code feels like it wants to be refactored. For example, the `fill` is created in both `if` branches in the same way. Also you're doing `data.where` with the same types of conditions and a fill value. Granted if you specify the `fill` for the float case you might have to make sure to do `np.float32(np.nan)` to preserve the 32-bit float dtype.

Eh maybe keep this the way it is. It is a lot of code in the middle of this method though.

Reply:
I would like to handle this separately.  I remember working on this when some of the AWIPS displays were not displaying correctly. I would want to test any changes here with the awips_tiled writer and that is a little more work.  If the values are flags, the fill value is retained in the data array attributes, otherwise it isn't.  I not completely convinced this had anything to do with how the data in AWIPS displayed, but I don't want to take the chance at this time.

Reply:
Sounds good. Understood. Leave it as it is.